In [1]:
from sklearn.tree import DecisionTreeClassifier 
from sklearn.model_selection import train_test_split 
from sklearn import metrics
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import pandas as pd

Random Forest: A classification machine learning method mainly used to predict catagorial data(in this case happiness)
Algorithm goes through n number of decision trees and summarizes the output of them in order to produce a final output.


Reading data, previewing

In [2]:
df_2019 = pd.read_csv('/Users/misaalsingh/Downloads/Full_Table.csv')
df_2019.columns = ['Year', 'Rank', 'Country', 'Happiness Score', 'GDP per Capita', 
                   'Social Support', 'Life Expectancy', 'Freedom', 'Generosity', 
                   'Corruption', 'Alcohol Consumption', 'Gas Emissions', 'Happy/IsNotHappy']
df_2019 = df_2019.fillna(6.165743055555556)
df_2019

Year  Rank                   Country  Happiness Score  GDP per Capita  \
0    2019     1                   Finland            7.769           1.340   
1    2019     2                   Denmark            7.600           1.383   
2    2019     3                    Norway            7.554           1.488   
3    2019     4                   Iceland            7.494           1.380   
4    2019     5               Netherlands            7.488           1.396   
..    ...   ...                       ...              ...             ...   
139  2019   152                    Rwanda            3.334           0.359   
140  2019   153                  Tanzania            3.231           0.476   
141  2019   154               Afghanistan            3.203           0.350   
142  2019   155  Central African Republic            3.083           0.026   
143  2019   156               South Sudan            2.853           0.306   

     Social Support  Life Expectancy  Freedom  Generosity  Corruption  \
0             1.587            0.986    0.596       0.153       0.393   
1             1.573            0.996    0.592       0.252       0.410   
2             1.582            1.028    0.603       0.271       0.341   
3             1.624            1.026    0.591       0.354       0.118   
4             1.522            0.999    0.557       0.322       0.298   
..              ...              ...      ...         ...         ...   
139           0.711            0.614    0.555       0.217       0.411   
140           0.885            0.499    0.417       0.276       0.147   
141           0.517            0.361    0.000       0.158       0.025   
142           0.000            0.105    0.225       0.235       0.035   
143           0.575            0.295    0.010       0.202       0.091   

     Alcohol Consumption  Gas Emissions Happy/IsNotHappy  
0                  10.78          51.30              Yes  
1                  10.26          34.23              Yes  
2                   7.41          18.43              Yes  
3                   9.12           2.11              Yes  
4                   9.61         152.52              Yes  
..                   ...            ...              ...  
139                 8.95           2.73               No  
140                11.27          75.68               No  
141                 0.21           7.59               No  
142                 2.38          14.31               No  
143                 6.16          10.28               No  

[144 rows x 13 columns]

Splitting the into two: test set and training set

Training set is used to train the random forest model and refine the calculations of the algorithm

Test set is used to asses the accuracy of the model

In [ ]:
col = ['GDP per Capita', 'Social Support', 'Life Expectancy', 'Freedom', 'Corruption',
                'Alcohol Consumption', 'Gas Emissions']
x = df_2019[col]
y = df_2019['Happy/IsNotHappy']

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=1)


Use the random forest classifier to create a forest object that is used to model the data. 

In [4]:
rf_model = RandomForestClassifier(n_estimators=50, max_features="auto", random_state=1)
rf_model.fit(x_train, y_train)

RandomForestClassifier(n_estimators=50, random_state=1)

In [6]:
pred = rf_model.predict(x_test)
pred

array(['No', 'Yes', 'No', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'No', 'No',
       'Yes', 'Yes', 'Yes', 'No', 'Yes', 'Yes', 'Yes', 'No', 'Yes', 'No',
       'Yes', 'Yes', 'No', 'Yes', 'No', 'No', 'No', 'Yes', 'Yes', 'Yes',
       'No', 'No', 'Yes', 'Yes', 'Yes', 'Yes', 'No', 'No', 'Yes', 'Yes',
       'No', 'Yes', 'Yes', 'No'], dtype=object)

In [7]:
accuracy_score(y_test, pred)

0.9772727272727273

After getting the final predictions we can see that the model is accurate 97.7% of the time

In [8]:
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

          No       1.00      0.94      0.97        18
         Yes       0.96      1.00      0.98        26

    accuracy                           0.98        44
   macro avg       0.98      0.97      0.98        44
weighted avg       0.98      0.98      0.98        44



Using a loop to output the weighting of each category towards its effect on Happiness

In [21]:
importances = rf_model.feature_importances_
columns = col
i = 0

while i < len(columns):
    print(f"The importance of feature '{columns[i]}' is {round(importances[i]*100,2)}%.")
    i+=1

The importance of feature 'GDP per Capita' is 18.19%.
The importance of feature 'Social Support' is 27.74%.
The importance of feature 'Life Expectancy' is 22.39%.
The importance of feature 'Freedom' is 9.38%.
The importance of feature 'Corruption' is 7.29%.
The importance of feature 'Alcohol Consumption' is 8.29%.
The importance of feature 'Gas Emissions' is 6.72%.
